csrc/index_compute.cpp

```diff
@@ -2138,6 +2138,77 @@ Val* Index::getProducerStridedIndices(
   }
 }
 
+namespace {
+
+bool shouldUseTensorIndexer(
+    const TensorView* producer,
+    const TensorView* consumer,
+    const std::unordered_set<ForLoop*>& rotated_loops) {
+  // Check if TensorIndexer is definitely required
+  auto is_tensor_indexer_required = [&]() -> bool {
+    bool is_producer_tma_op = producer->definition() != nullptr &&
+        producer->definition()->isA<LoadStoreOp>() &&
+        ir_utils::isCpAsyncBulkLoad(producer->definition());
+    bool is_consumer_tma_op = consumer->definition() != nullptr &&
+        consumer->definition()->isA<LoadStoreOp>() &&
+        ir_utils::isCpAsyncBulkLoad(consumer->definition());
+
+    return !ir_utils::hasRootToLoopLinearTransformations(producer) ||
+        (consumer->definition()->isA<MmaOp>() &&
+         isHopper(consumer->definition()->as<MmaOp>()->macro())) ||
+        is_producer_tma_op || is_consumer_tma_op ||
+        GpuLower::current()->tmemInfo().hasTMemTensor();
+  };
+
+  // Check if TensorIndexer is supported.
+  auto is_tensor_indexer_supported = [&](bool assert) -> bool {
+    bool is_producer_ldmatrix_op = producer->definition() != nullptr &&
+        producer->definition()->isA<LoadStoreOp>() &&
+        producer->definition()->as<LoadStoreOp>()->opType() ==
+            LoadStoreOpType::LdMatrix;
+    bool is_producer_stmatrix_op_with_no_alloc_domain =
+        producer->definition() != nullptr &&
+        producer->definition()->isA<LoadStoreOp>() &&
+        producer->definition()->as<LoadStoreOp>()->opType() ==
+            LoadStoreOpType::StMatrix &&
+        !producer->hasAllocation();
```

Comment:
In the case of `stmatrix`, it seems to work as long as `producer` has a set allocation domain.